cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,417 @@
+	return []string{agentPackageKey}, nil
+}
+
+func (p packageManager) PackageState(packageName string) (state types.PackageState, err error) {
+	if packageName == agentPackageKey {
+		return types.PackageState{
+			Exists:  true,
+			Type:    protobufs.PackageType_PackageType_TopLevel,
+			Hash:    p.topLevelHash,
+			Version: p.topLevelVersion,
+		}, nil
+	}
+
+	return types.PackageState{
+		Exists: false,
+	}, nil
+}
+
+func (p *packageManager) SetPackageState(packageName string, state types.PackageState) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package %q does not exist", packageName)
+	}
+
+	if !state.Exists {
+		return fmt.Errorf("agent package must be marked as existing")
+	}
+
+	if state.Type != protobufs.PackageType_PackageType_TopLevel {
+		return fmt.Errorf("agent package must be marked as top level")
+	}
+
+	p.topLevelHash = state.Hash
+	p.topLevelVersion = state.Version
+
+	return nil
+}
+
+func (packageManager) CreatePackage(packageName string, _ protobufs.PackageType) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("only agent package is supported")
+	}
+
+	return fmt.Errorf("agent package already exists")
+}
+
+func (p *packageManager) FileContentHash(packageName string) ([]byte, error) {
+	if packageName != agentPackageKey {
+		return nil, nil
+	}
+
+	return p.topLevelHash, nil
+}
+
+func (p *packageManager) UpdateContent(ctx context.Context, packageName string, data io.Reader, contentHash, signature []byte) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package does not exist")
+	}
+
+	by, err := io.ReadAll(data)
+	if err != nil {
+		return fmt.Errorf("read package bytes: %w", err)
+	}
+
+	err = verifyPackageIntegrity(by, contentHash)
+	if err != nil {
+		return fmt.Errorf("could not verify package integrity: %w", err)
+	}
+
+	b64Cert, b64Signature, err := parsePackageSignature(signature)
+	if err != nil {
+		return fmt.Errorf("could not parse package signature: %w", err)
+	}
+
+	err = verifyPackageSignature(ctx, p.checkOpts, by, b64Cert, b64Signature)
+	if err != nil {
+		return fmt.Errorf("could not verify package signature: %w", err)
+	}
+
+	// overwrite agent process
+	startAgent, err := p.am.stopAgentProcess(ctx)
+	if err != nil {
+		return fmt.Errorf("stop agent process: %w", err)
+	}
+
+	// We always want to start the agent process again, even if we fail to write the agent file
+	defer close(startAgent)
+
+	// Create a backup in case we fail to write the agent
+	agentBackupPath := filepath.Join(p.storageDir, "collector.bak")
+	backupFile, err := os.OpenFile(agentBackupPath, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return fmt.Errorf("open backup file: %w", err)
+	}
+	defer backupFile.Close()
+
+	agentFile, err := os.OpenFile(p.agentPath, os.O_RDWR, 0700)
+	if err != nil {
+		return fmt.Errorf("open file: %w", err)
+	}
+	defer agentFile.Close()
+
+	// Copy to backup
+	_, err = io.Copy(backupFile, agentFile)
```

Comment:
A possible alternate instead of copying the files is to rename them. Delete backup, rename original to backup, write to the original name. Skips the copying and is likely faster.